accepted/0000-ubuntu-sudo.md

```diff
@@ -0,0 +1,63 @@
+- Feature Name: automated ubuntu onboarding without need to manually configure sudo
+- Start Date: (2020-01-27)
+- RFC PR:
+
+# Summary
+[summary]: #summary
+
+As Ubuntu clients have no root users enabled by default, we tell the user to manually use visudo to enable a user for Python (i. e. for Salt)
+https://opensource.suse.com/doc-susemanager/suse-manager/develop/client-configuration/clients-ubuntu.html#_root_access
+
+This proposal tries to avoid the need to manually configure sudo before being able to bootstrap Ubuntu clients.
+
+# Motivation
+[motivation]: #motivation
+
+- Why are we doing this?
+
+Better user experience, make it easier to onboard Ubuntu minions
+
+- What use cases does it support?
+
+Onboarding Ubuntu minions
+
+- What is the expected outcome?
+
+Ubuntu clients should be as easy to onboard as any other client. This means it should be sufficient to just enter the root password in the SUSE Manager UI without also manually configuring the to-be-onboarded Ubuntu minion.
+
+# Detailed design
+[design]: #detailed-design
+
+When onboarding a minion via ssh, SUSE Manager asks for a user and password with sufficient privileges. On non-Ubuntu systems this typically is the root user. On Ubuntu systems, there typically is just an ordinary user, but this user can do almost anything via sudo. This user is referred to as "ubuntuser" in this document. Unfortunately it is necessary to supply the user's password to sudo. So currently the SUSE Manager documentation asks the admin to enable some special commands in sudo to be executed even without providing the password. The goal is to do this modification directly from SUSE Manager, so there is no need to manually prepare the system for onboarding.
+
+So the actual workflow would be like this:
+
+SUSE Manager UI asks for user and password like with any other client. The password needs to get stored in some local temporary file, so it can be transferred to the minion without being visible in the process list:
+
+1. umask 0377 && echo $PASSWORD > /run/usr/$UID/pass.txt
+2. scp -p /run/usr/$UID/pass.txt ubuntuser@ubuntu-minion.target.system:/run/usr/$UID/pass.txt
```

Comment:
I understand the idea behind those commands, but I wonder how we would actually implement something like that. At the moment Java code does not shell out to ssh, rather we call the Salt API to drive Salt SSH (to apply some states).

How would you imagine sneaking in those commands (or something else to a similar effect) via Salt SSH, and how would Java code drive that?

Reply:
Oh, I was not aware of the actual current implementation. But isn't there some mechanism we can use? How is the onboarding performed in cases where no salt agent is installed? So far this onboarding mechanism was a black box to me, maybe Matei can comment as he implemented it initially.

Reply:
> I understand the idea behind those commands, but I wonder how we would actually implement something like that. At the moment Java code does not shell out to ssh, rather we call the Salt API to drive Salt SSH (to apply some states).
> 
> How would you imagine sneaking in those commands (or something else to a similar effect) via Salt SSH, and how would Java code drive that?

I'm not sure I understand what the problem is.

There are a number of Java libraries providing an SSH implementation. If we are not using any of them, we should pick one and add it to our dependencies. E. g. here is an example of what we'd need, implemented with JSch (pure Java library, BSD-licensed)
https://medium.com/@ldclakmal/scp-with-java-b7b7dbcdbc85
http://www.jcraft.com/jsch/

Reply:
What we do from the UI is to use salt-ssh (aka masterless Salt) to apply a state which does the same things a regular bootstrap script does.

So when it comes to minions, Java code knows nothing about the SSH layer, all it does is to delegate to Salt via the usual Salt APIs. I am not sure we want to introduce anything jsch based on the Java side to go straight to the minion before handing control over Salt SSH - it sounds like a layering violation to me.

@mateiw would it be possible to still use Salt SSH to apply a first state that configures sudo, followed by our existing states to bootstrap everything else?

@paususe the class you mentioned implements part of the SSH Push client protocol, aka one of the possible modes of the traditional stack. Specifically traditional cliends have a (pull-based) agent service by default, and on top of that we developed the so-called "SSH Push" mode to deal with those cases when an active agent is not acceptable, but plain SSH is an option.

Reply:
> What we do from the UI is to use [salt-ssh](https://docs.saltstack.com/en/latest/topics/ssh/) (aka masterless Salt) to apply a state which does the same things a regular bootstrap script does.
> 
> So when it comes to minions, Java code knows nothing about the SSH layer, all it does is to delegate to Salt via the usual Salt APIs. I am not sure we want to introduce anything jsch based on the Java side to go straight to the minion before handing control over Salt SSH - it sounds like a layering violation to me.
> 
> @mateiw would it be possible to still use Salt SSH to apply a first state that configures sudo, followed by our existing states to bootstrap everything else?

What would be the advantage of using Salt SSH or JSch? In the end, we need no Salt feature here.
 
> @paususe the class you mentioned implements part of the SSH Push client protocol, aka one of the possible modes of the traditional stack. Specifically traditional cliends have a (pull-based) agent service by default, and on top of that we developed the so-called "SSH Push" mode to deal with those cases when an active agent is not acceptable, but plain SSH is an option.

Thanks for the information but I only mentioned that class because it already uses JSch, so it's not like we are introducing a new dependency here.

Reply:
> What would be the advantage of using Salt SSH or JSch? In the end, we need no Salt feature here.

Avoiding to add yet another new completely separate code path to achieve the same goal: running an arbitrary command via SSH.

Reply:
> > What would be the advantage of using Salt SSH or JSch? In the end, we need no Salt feature here.
> 
> Avoiding to add yet another new completely separate code path to achieve the same goal: running an arbitrary command via SSH.

What about efficiency?

JSCh => salt-ssh (involves copying Salt) => run 1-liner shell script that requires nothing provided by Salt

vs

JSch => run 1-liner shell script that requires nothing provided by Salt

I don't understand using salt-ssh, which we do not need at all, for such a simple operation.

Reply:
We are using salt-ssh anyway for the rest of the onboarding process, so performance wise copying and all is a sunk cost FMPOV.

Reply:
According to the salt-ssh docs, NOPASSWD in sudoers is required for salt-ssh to work:
https://docs.saltstack.com/en/latest/topics/ssh/

That 1-liner is what adds NOPASSWD to the sudoers file. 

Will we have a chicken and egg problem here if we try to run it with salt-ssh? 

Or can we use "salt-ssh - r" , set "sudo: False" in the roster file and then make sudo part of the command (ie what the RFC contains)? 

Also, I can't remember at the moment if logging out if required to enable the changes in the sudoers file (and I cannot test atm)

Reply:
That was what I had in mind to research about, thanks for starting it.

I plan to give the topic an in depth look from this angle as soon as time allows.

Reply:
Why you? Is this scalability-related?

Reply:
It's not in my area, but I do try to take a look at all incoming RFCs in general and help where I can :-)

Reply:
@meaksh Could you please shed some light?